### PR TITLE
Razor snippets

### PR DESCRIPTION
**Pull request by dibarbet**
Related to https://github.com/dotnet/razor-tooling/pull/6012/

Implements support for built in snippets for Razor LSP using inline completions.  This reads the XML snippet information, applies function results and then converts into an LSP formatted snippet.

While this should work for many XML snippets including custom ones, there are edge cases that I haven't tackled yet (no selection support, no add imports/references support).  As such I've restricted this to built in snippets .  Support for the remaining will require further iteration on both the lsp client portion and server world.

I will additionally have a followup PR to cache the snippets as we parse from XML.
razor_snippets

### CODE REVIEW

File: src/Features/CSharp/Portable/Snippets/CSharpSnippetFunctionService.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+[ExportLanguageService(typeof(SnippetFunctionService), LanguageNames.CSharp), Shared]
+internal class CSharpSnippetFunctionService : SnippetFunctionService
```
Comment (dibarbet) [inline src/Features/CSharp/Portable/Snippets/CSharpSnippetFunctionService.cs:17]: extracted this code from the c# snippet functions
https://github.com/dotnet/roslyn/tree/main/src/VisualStudio/CSharp/Impl/Snippets/SnippetFunctions

File: src/Features/Core/Portable/Snippets/SnippetFunctionService.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis;
+internal abstract class SnippetFunctionService : ILanguageService
```
Comment (dibarbet) [inline src/Features/Core/Portable/Snippets/SnippetFunctionService.cs:17]: extracted from https://github.com/dotnet/roslyn/tree/main/src/VisualStudio/Core/Def/Implementation/Snippets/SnippetFunctions

File: src/Features/Core/Portable/Snippets/SnippetUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.LanguageServices;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis;
+internal class SnippetUtilities
+{
+    public static bool TryGetWordOnLeft(int position, SourceText currentText, ISyntaxFactsService syntaxFactsService, [NotNullWhen(true)] out TextSpan? wordSpan)
```
Comment (dibarbet) [inline src/Features/Core/Portable/Snippets/SnippetUtilities.cs:12]: extracted from https://github.com/dotnet/roslyn/blob/b0f460c295f1d57c5b217d286958595c57e2a45b/src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetCommandHandler.cs#L295

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.XmlSnippetParsing.cs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Xml;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.InlineCompletions;
+
+internal partial class InlineCompletionsHandler
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.XmlSnippetParsing.cs:18]: code here was extracted from similar editor side code that reads snippet xml files and modified a bit to support our use case.

If we need to support XML based snippets long term, it would be nice if an API existed on the editor side to retrieve these.  We would have to still do some conversion since some snippets are context dependent, but reading and parsing the XML could be left to the editor.

That being said, semantic snippets will probably make this irrelevant for built in c# snippets, and the editor is re-thinking the snippet experience so I don't believe it's worth doing anything right now.

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.XmlSnippetParsing.cs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+                _tokens.Add(new ExpansionField(
+                    CodeSnippet.GetElementInnerText(declarationElement, "ID"),
+                    CodeSnippet.GetElementInnerText(declarationElement, "Default") ?? " ",
+                    functionName,
+                    functionParam,
+                    editableAttribute == null || string.Equals(editableAttribute.Value, "true", StringComparison.Ordinal) || string.Equals(editableAttribute.Value, "1", StringComparison.Ordinal)));
+            }
+        }
+
+        public ParsedXmlSnippet? Parse()
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.XmlSnippetParsing.cs:170]: much of the parsing logic remains the same as the battle-tested editor code, the major change is to convert the snippet into 'SnippetParts' instead of just a pure string with default values.

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                    {
+                        return this;
+                    }
+
+                    return this with { DefaultText = className };
+                case "GenerateSwitchCases":
+                    // Generate switch cases requires a multi-step snippet interaction, where the snippet is inserted first then the
+                    // client calls back to on commit so that we can generate the cases for the specified switch value.
+                    // This is not yet supported via LSP.
+                    return this;
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs:73]: Didn't realize this is a two step process at the time of writing the LSP api :(
Future work here is to figure out a way for LSP to call back on the 'commit' of a tab stop and ask the server for more edits.

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.InlineCompletions;
+
+/// <summary>
+/// Supports built in legacy snippets for razor scenarios.
+/// </summary>
+[ExportRoslynLanguagesLspRequestHandlerProvider, Shared]
+[ProvidesMethod(VSInternalMethods.TextDocumentInlineCompletionName)]
+internal partial class InlineCompletionsHandler : AbstractStatelessRequestHandler<VSInternalInlineCompletionRequest, VSInternalInlineCompletionList>
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:32]: the meat and potatoes of the actual new logic (that wasn't copied from the editor)

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+        // We currently only support snippet expansions, the others require selection support which is not applicable here.
+        if (!matchingSnippet.SnippetTypes.Contains("Expansion", StringComparer.OrdinalIgnoreCase))
+        {
+            return new VSInternalInlineCompletionList();
+        }
+
+        var expansion = new ExpansionTemplate(matchingSnippet);
+
+        // Parse the snippet XML
+        var parsedSnippet = expansion.Parse();
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:107]: the parsed snippet needs to be cached.  Am working on that, but will do as a separate PR to reduce the size here

Note that we cannot cache the actual LSP snippet, as it is context dependent (class names / formatting), but we can cache the snippet parts we read from XML.

File: src/VisualStudio/CSharp/Impl/Snippets/SnippetFunctions/SnippetFunctionClassName.cs [hunk 9]
```diff
@@ -1,42 +0,0 @@
...
-using System.Threading;
-using Microsoft.CodeAnalysis;
-using Microsoft.CodeAnalysis.CSharp.Syntax;
-using Microsoft.CodeAnalysis.Shared.Extensions;
-using Microsoft.VisualStudio.LanguageServices.Implementation.Snippets;
-using Microsoft.VisualStudio.Text;
-
-namespace Microsoft.VisualStudio.LanguageServices.CSharp.Snippets.SnippetFunctions
-{
-    internal sealed class SnippetFunctionClassName : AbstractSnippetFunctionClassName
```
Comment (dibarbet) [inline src/VisualStudio/CSharp/Impl/Snippets/SnippetFunctions/SnippetFunctionClassName.cs:16]: the logic here can now live in a single impl for each language since the language specific logic is now in the snippetfunctionservice (and children)

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+[ExportRoslynLanguagesLspRequestHandlerProvider, Shared]
+[ProvidesMethod(VSInternalMethods.TextDocumentInlineCompletionName)]
+internal partial class InlineCompletionsHandler : AbstractStatelessRequestHandler<VSInternalInlineCompletionRequest, VSInternalInlineCompletionList>
+{
+    /// <summary>
+    /// The set of built in snippets from, typically found in
+    /// C:\Program Files\Microsoft Visual Studio\2022\VS_INSTANCE\VC#\Snippets\1033\Visual C#
+    /// These are currently the only snippets supported.
+    /// </summary>
+    public static ImmutableHashSet<string> BuiltInSnippets = ImmutableHashSet.Create(
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:39]: I've limited these to built in snippets for now.  While the parsing could probably handle most custom snippets, there are a few things that would not work, namely snippets with additional references (requires lsp response to support additional text edits) and selection (currently snippets are only invocable via tab-tab, so selection does not make sense in that context).

Depending on the direction that the editor takes xml snippets and where we go with semantic snippets, we can potentially open this up to more with a few improvements on the lsp api.

File: src/EditorFeatures/Core/Implementation/LanguageServer/VisualStudioInProcLanguageServer.cs [hunk 11]
```diff
@@ -144,6 +144,15 @@ void PublishOpenFileDiagnostics()
                 renameParams, _clientCapabilities, ClientName, cancellationToken);
         }
 
+        [JsonRpcMethod(VSInternalMethods.TextDocumentInlineCompletionName, UseSingleObjectParameterDeserialization = true)]
+        public Task<VSInternalInlineCompletionList> GetInlineCompletionsAsync(VSInternalInlineCompletionRequest request, CancellationToken cancellationToken)
```
Comment (JoeRobich) [inline src/EditorFeatures/Core/Implementation/LanguageServer/VisualStudioInProcLanguageServer.cs:148]:
```suggestion
        public Task<VSInternalInlineCompletionList?> GetInlineCompletionsAsync(VSInternalInlineCompletionRequest request, CancellationToken cancellationToken)
```

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs (hunk 6 above)
Comment (NTaylorMullen) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs:73]: Interesting, is this also handled in the new semantic snippets?

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.InlineCompletions;
+
+internal partial class InlineCompletionsHandler
```
Comment (NTaylorMullen) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs:14]: Do you typically utilize partial classes heavily in Roslyn? They typically fall over in some source linked debug scenarios + make understanding implementation details a bit more tedious. Would it be awful to pull all the XML snippet tech into a service that the primary InlineCompletionsHandler utilizes?

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs (hunk 8 above)
Comment (NTaylorMullen) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:107]: We also may want to build specific error handling around this given it's reading user content. Also might make sense to do the same with `RetrieveSnippetFromXml` above given that reads content from disk

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+            if (part is SnippetFunctionPart functionPart)
+            {
+                // To avoid a bunch of document changes and re-parsing, we always calculate the snippet function result
+                // against the document with the default snippet text applied to it instead of with each incremental function result.
+                // So we need to remember the index into the original document.
+                part = await functionPart.WithSnippetFunctionResultAsync(documentWithDefaultSnippet, new TextSpan(locationInDefaultSnippet, part.DefaultText.Length), cancellationToken).ConfigureAwait(false);
+            }
+
+            // Only store spans for editable fields or the cursor location, we don't need to get back to anything else.
+            if (part is SnippetFieldPart fieldPart && fieldPart.EditIndex != null)
```
Comment (NTaylorMullen) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:245]: Can the above function part translate the part into a `SnippetFieldPart` or should this be an `else if`

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+        if (!SnippetUtilities.TryGetWordOnLeft(position, sourceText, syntaxFactsService, out var wordOnLeft))
+        {
+            return new VSInternalInlineCompletionList();
+        }
+
+        // Find the snippet with shortcut text that matches the typed word.
+        var wordText = sourceText.GetSubText(wordOnLeft.Value).ToString();
+        if (!BuiltInSnippets.Contains(wordText, StringComparer.OrdinalIgnoreCase))
+        {
+            return new VSInternalInlineCompletionList();
```
Comment (NTaylorMullen) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:87]: Ah does the API not support returning `null`? If it does it might be clear to return that here as a way to very clearly say "we have nothing". Same with above

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+        }
+
+        var matchingSnippetInfo = snippetInfo.Single(s => wordText.Equals(s.Shortcut, StringComparison.OrdinalIgnoreCase));
```
Comment (NTaylorMullen) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:90]: Would this throw if a user had a custom snippet with a similar shortcut under a different file name? I'm imagining a customer who was playing around with custom snippets and copy & pasted a pre-existing snippet

**Review by NTaylorMullen — COMMENTED** [review]
This is super exciting and great work! I imagine the path you're taking here would also be used by our friends over on the XAML language service side too given it could potentially enable custom snippets too. Quite flexible.

I had some overarching concerns around the snippet parsing tech but reaching out offline about that.

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs (hunk 6 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs:73]: semantic snippets would have to support it in some manner - the cases can only be known when the field is committed

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs (hunk 13 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:245]: a function part is a `SnippetFieldPart`

this is not an else if since they're doing two different actions
A function is a field, but to get the proper value to insert we need to apply the function based on the snippet default text (what happens above).
Here, we store spans for editable fields.  Note that a function field could also be an editable field so this needs to be its own if.

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs (hunk 15 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:90]: True - will switch to first as I don't have a better solution.

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs (hunk 12 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs:14]: partial classes are not uncommon in roslyn.  It sounds like source link debug scenarios need to be more resilient ;)

I don't mind either way really

**Comment by mgoertz-msft** [comment]
> This is super exciting and great work! I imagine the path you're taking here would also be used by our friends over on the XAML language service side too given it could potentially enable custom snippets too. Quite flexible.

Totally agree! I haven't fully wrapped my mind around how to utilize this for custom snippets in XAML but I'm excited to see that the foundation has been laid.

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs (hunk 8 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:107]: I think I'm OK with the amount of error handling in there right now.  currently we do some basic checks ot make sure the file exists.  But otherwise if there are malformed snippets, imho it is ok to let those exceptions bubble up to the queue and let them be automatically reported and logged.

Happy to hear other thoughts though.

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:107]: I did find a few cases I should probably handle actually, so will update

**Review by NTaylorMullen — APPROVED** [review]
:shipit: It's happpppennnningggg!@!!!
